---
title: RNA-seq原始数据处理
author: Xiaotao Shen
date: '2021-06-30'
slug: 
categories:
  - Linux
tags:
  - Blog
  - Chinese
image:
  caption: ''
  focal_point: ''
output:
  blogdown::html_page:
    toc: true
    number_sections: true
---

<script src="index_files/header-attrs/header-attrs.js"></script>

<div id="TOC">
<ul>
<li><a href="#rna-seq原始数据"><span class="toc-section-number">1</span> RNA-seq原始数据</a></li>
<li><a href="#数据整理"><span class="toc-section-number">2</span> 数据整理</a>
<ul>
<li><a href="#raw_data"><span class="toc-section-number">2.1</span> raw_data</a></li>
<li><a href="#qc_results"><span class="toc-section-number">2.2</span> qc_results</a></li>
<li><a href="#clean_data"><span class="toc-section-number">2.3</span> clean_data</a></li>
<li><a href="#reference_geonome"><span class="toc-section-number">2.4</span> reference_geonome</a></li>
<li><a href="#aligned"><span class="toc-section-number">2.5</span> aligned</a></li>
</ul></li>
<li><a href="#质量控制"><span class="toc-section-number">3</span> 质量控制</a>
<ul>
<li><a href="#fastqc"><span class="toc-section-number">3.1</span> fastQC</a></li>
<li><a href="#multiqc"><span class="toc-section-number">3.2</span> multiqc</a></li>
</ul></li>
<li><a href="#使用trim-galore去除低质量的reads和adaptor"><span class="toc-section-number">4</span> 使用trim-galore去除低质量的reads和adaptor</a></li>
<li><a href="#序列比对"><span class="toc-section-number">5</span> 序列比对</a></li>
<li><a href="#samtools将sam转换为bam"><span class="toc-section-number">6</span> samtools将sam转换为bam</a></li>
</ul>
</div>

<div id="rna-seq原始数据" class="section level1" number="1">
<h1><span class="header-section-number">1</span> RNA-seq原始数据</h1>
<p>参考:<a href="https://mp.weixin.qq.com/s?__biz=MzA4NDAzODkzMA==&amp;mid=2651272899&amp;idx=1&amp;sn=6779b2bd21f3b607a08227d31c7212c6&amp;chksm=841ed2beb3695ba8bee23563c28caa005447b2298785719964732b16cafe3a15d7d4937b95c1&amp;scene=21#wechat_redirect" class="uri">https://mp.weixin.qq.com/s?__biz=MzA4NDAzODkzMA==&amp;mid=2651272899&amp;idx=1&amp;sn=6779b2bd21f3b607a08227d31c7212c6&amp;chksm=841ed2beb3695ba8bee23563c28caa005447b2298785719964732b16cafe3a15d7d4937b95c1&amp;scene=21#wechat_redirect</a></p>
<p>原始数据格式为fastq,为本文文件.用来保存生物序列.每一个样本一个fastq文件,每个序列用四行信息记录.</p>
</div>
<div id="数据整理" class="section level1" number="2">
<h1><span class="header-section-number">2</span> 数据整理</h1>
<p>一般需要将不同数据放在不同文件夹下.</p>
<div id="raw_data" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> raw_data</h2>
<p>原始数据.fastq或者fastq.gz格式.</p>
</div>
<div id="qc_results" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> qc_results</h2>
<p>用来存放质量控制得到的数据.</p>
</div>
<div id="clean_data" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> clean_data</h2>
<p>用来存放<code>trim_galore</code>清洗之后的数据.</p>
</div>
<div id="reference_geonome" class="section level2" number="2.4">
<h2><span class="header-section-number">2.4</span> reference_geonome</h2>
<p>用来存放参考基因组.</p>
</div>
<div id="aligned" class="section level2" number="2.5">
<h2><span class="header-section-number">2.5</span> aligned</h2>
<p>用来存放比对之后的数据.</p>
</div>
</div>
<div id="质量控制" class="section level1" number="3">
<h1><span class="header-section-number">3</span> 质量控制</h1>
<p>使用<code>fastQC</code>和<code>multiqc</code>对测序质量进行评价.</p>
<div id="fastqc" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> fastQC</h2>
<p>首先使用fastQC对每个样品的测序质量进行评估.</p>
<pre><code>fastqc -o &lt;output dir&gt; &lt;seqfile1,seqfile2..&gt;</code></pre>
<p>-o:输出的路径</p>
<p>&lt;seqfile1,seqfile2..&gt;:要进行评估的原始数据.</p>
<p>将路径设置到最外面.</p>
<p>然后输入下面代码:</p>
<pre><code>fastqc -o qc_results *.fastq.gz</code></pre>
</div>
<div id="multiqc" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> multiqc</h2>
<p>使用multiqc将fastqc对每个测序样品的结果进行汇总.</p>
<pre><code>multiqc *fastqc.zip --pdf</code></pre>
<p>fastQC的结果是fastqc.zip格式.</p>
<p>需要先将路径设置到<code>qc_results</code>中.然后运行该命令.</p>
</div>
</div>
<div id="使用trim-galore去除低质量的reads和adaptor" class="section level1" number="4">
<h1><span class="header-section-number">4</span> 使用trim-galore去除低质量的reads和adaptor</h1>
<p>处理单个样本可以使用下面命令.</p>
<pre><code>trim_galore -output_dir clean_data --paired --length 75 --quality 25 --stringency 5 /raw_data/*fastq.gz</code></pre>
<p>批量多核处理:</p>
<p>首先设置路径到<code>raw_data</code></p>
<pre><code>ls|grep R1_001.fastq.gz&gt;gz1
ls|grep R2_001.fastq.gz&gt;gz2
paste gz1 gz2&gt;config
cat config</code></pre>
<p>然后创建一个<code>trim.sh</code>文件:</p>
<pre><code>touch trim.sh</code></pre>
<p>vim打开.</p>
<pre><code>vim trim.sh</code></pre>
<p>按esc,然后<code>:i</code>进入插入模式,写入下面的代码:</p>
<pre><code>dir=/home/kelly/wesproject/clean_data/
cat config |while read id
do
      arr=${id}
      fq1=${arr[0]}
      fq2=${arr[1]}
      nohup trim_galore -q 25 --phred33 --length 36 -e 0.1 --stringency 3 --paired -o $dir $fq1 $fq2 &amp;
done</code></pre>
<p>其中dir是clean_data的绝对路径.</p>
<p>然后按esc,然后<code>:wq</code>保存退出.</p>
<p>然后运行该文件.</p>
<pre><code>shell trim.sh</code></pre>
<p>然后就会在clean_data中生成数据.</p>
<p>进入该文件夹下.可以查看文件的数据.</p>
<pre><code>ls -lht | grep val | wc -l</code></pre>
<p>这种写法<code>|</code>跟R中的<code>%&gt;%</code>类似.</p>
<p>这里面,<code>wc -l</code>是计数的.</p>
</div>
<div id="序列比对" class="section level1" number="5">
<h1><span class="header-section-number">5</span> 序列比对</h1>
<p>我们使用<code>hisat</code>进行序列比对.</p>
<p>下载参考基因组:</p>
<p>路径设置到想要存放路径的地方.</p>
<pre><code>wget ftp://ftp.ccb.jhu.edu/pub/infphilo/hisat2/data/hg38.tar.gz</code></pre>
<p>这里我们下载最近的hg38参考基因组.</p>
<p>hisat2使用.</p>
<pre><code>hisat2 -p 6 -x &lt;dir of index of genome&gt; -1 seq_val_1.fq.gz -2 seq_val_2.fq.gz -S  tem.hisat2.sam</code></pre>
<p>-p: 处理核心数.</p>
<p>-x: 参考基因组存放位置.最后需要写上<code>geome</code>.</p>
<p>-1: 两端测序的第一个文件.</p>
<p>-2: 两端测序的第二个文件.</p>
<p>-S: 生成的sam格式数据的名字.</p>
<p>具体例子:</p>
<pre><code>hisat2 --dta -t -p 8 -x ./reference_genome/index/hg38/genome -1 ./clean_data/iPOP_MC_PBMC_RNAseq_1_S1_L001_R1_001_val_1.fq.gz -2 ./clean_data/iPOP_MC_PBMC_RNAseq_1_S1_L001_R2_001_val_2.fq.gz -S ./aligned/iPOP_MC_PBMC_RNAseq_1_S1_L001_R2_001_val_2.sam</code></pre>
<p>路径需要在整个project root路径.</p>
<p>注意参考基因组的路径写法,其实我们的参考基因组就存放在<code>./reference_genome/index/hg38/</code>文件夹下.genome一定要加上去.</p>
<p>写一个循环进行批次处理:</p>
<pre><code>for i in iPOP_MC_PBMC_RNAseq_1_S1_L001 iPOP_MC_PBMC_RNAseq_10_S10_L001 iPOP_MC_PBMC_RNAseq_11_S11_L001 iPOP_MC_PBMC_RNAseq_12_S12_L001 iPOP_MC_PBMC_RNAseq_13_S13_L001 iPOP_MC_PBMC_RNAseq_14_S14_L001 iPOP_MC_PBMC_RNAseq_15_S15_L001 iPOP_MC_PBMC_RNAseq_16_S16_L001 iPOP_MC_PBMC_RNAseq_17_S17_L001 iPOP_MC_PBMC_RNAseq_18_S18_L001 iPOP_MC_PBMC_RNAseq_19_S19_L001 iPOP_MC_PBMC_RNAseq_2_S2_L001 iPOP_MC_PBMC_RNAseq_20_S20_L001 iPOP_MC_PBMC_RNAseq_21_S21_L001 iPOP_MC_PBMC_RNAseq_22_S22_L001 iPOP_MC_PBMC_RNAseq_23_S23_L001 iPOP_MC_PBMC_RNAseq_24_S24_L001 iPOP_MC_PBMC_RNAseq_25_S25_L001 iPOP_MC_PBMC_RNAseq_26_S26_L001 iPOP_MC_PBMC_RNAseq_27_S27_L001 iPOP_MC_PBMC_RNAseq_28_S28_L001 iPOP_MC_PBMC_RNAseq_29_S29_L001 iPOP_MC_PBMC_RNAseq_3_S3_L001 iPOP_MC_PBMC_RNAseq_30_S30_L001 iPOP_MC_PBMC_RNAseq_31_S31_L001 iPOP_MC_PBMC_RNAseq_32_S32_L001 iPOP_MC_PBMC_RNAseq_33_S33_L001 iPOP_MC_PBMC_RNAseq_34_S34_L001 iPOP_MC_PBMC_RNAseq_35_S35_L001 iPOP_MC_PBMC_RNAseq_36_S36_L001 iPOP_MC_PBMC_RNAseq_37_S37_L001 iPOP_MC_PBMC_RNAseq_38_S38_L001 iPOP_MC_PBMC_RNAseq_39_S39_L001 iPOP_MC_PBMC_RNAseq_4_S4_L001 iPOP_MC_PBMC_RNAseq_40_S40_L001 iPOP_MC_PBMC_RNAseq_41_S41_L001 iPOP_MC_PBMC_RNAseq_42_S42_L001 iPOP_MC_PBMC_RNAseq_43_S43_L001 iPOP_MC_PBMC_RNAseq_44_S44_L001 iPOP_MC_PBMC_RNAseq_45_S45_L001 iPOP_MC_PBMC_RNAseq_46_S46_L001 iPOP_MC_PBMC_RNAseq_47_S47_L001 iPOP_MC_PBMC_RNAseq_5_S5_L001 iPOP_MC_PBMC_RNAseq_6_S6_L001 iPOP_MC_PBMC_RNAseq_7_S7_L001 iPOP_MC_PBMC_RNAseq_8_S8_L001 iPOP_MC_PBMC_RNAseq_9_S9_L001 Undetermined_S0_L001
do 
hisat2 --dta -t -p 15 -x ./reference_genome/index/hg38/genome \
-1 ./clean_data/&quot;$i&quot;_R1_001_val_1.fq.gz \
-2 ./clean_data/&quot;$i&quot;_R2_001_val_2.fq.gz \
-S ./aligned/&quot;$i&quot;.sam; done</code></pre>
</div>
<div id="samtools将sam转换为bam" class="section level1" number="6">
<h1><span class="header-section-number">6</span> samtools将sam转换为bam</h1>
<p>使用samtools将得到的sam格式数据转换为bam格式,并且进行sort.</p>
<p>单个转换:</p>
<pre><code>samtools view -S iPOP_MC_PBMC_RNAseq_10_S10_L001.sam -b &gt; iPOP_MC_PBMC_RNAseq_10_S10_L001.bam</code></pre>
<p>批次转换:</p>
<pre><code>for i in iPOP_MC_PBMC_RNAseq_11_S11_L001 iPOP_MC_PBMC_RNAseq_12_S12_L001 iPOP_MC_PBMC_RNAseq_13_S13_L001 iPOP_MC_PBMC_RNAseq_14_S14_L001 iPOP_MC_PBMC_RNAseq_15_S15_L001 iPOP_MC_PBMC_RNAseq_16_S16_L001 iPOP_MC_PBMC_RNAseq_17_S17_L001 iPOP_MC_PBMC_RNAseq_18_S18_L001 iPOP_MC_PBMC_RNAseq_19_S19_L001 iPOP_MC_PBMC_RNAseq_2_S2_L001 iPOP_MC_PBMC_RNAseq_20_S20_L001 iPOP_MC_PBMC_RNAseq_21_S21_L001 iPOP_MC_PBMC_RNAseq_22_S22_L001 iPOP_MC_PBMC_RNAseq_23_S23_L001 iPOP_MC_PBMC_RNAseq_24_S24_L001 iPOP_MC_PBMC_RNAseq_25_S25_L001 iPOP_MC_PBMC_RNAseq_26_S26_L001 iPOP_MC_PBMC_RNAseq_27_S27_L001 iPOP_MC_PBMC_RNAseq_28_S28_L001 iPOP_MC_PBMC_RNAseq_29_S29_L001 iPOP_MC_PBMC_RNAseq_3_S3_L001 iPOP_MC_PBMC_RNAseq_30_S30_L001 iPOP_MC_PBMC_RNAseq_31_S31_L001 iPOP_MC_PBMC_RNAseq_32_S32_L001 iPOP_MC_PBMC_RNAseq_33_S33_L001 iPOP_MC_PBMC_RNAseq_34_S34_L001 iPOP_MC_PBMC_RNAseq_35_S35_L001 iPOP_MC_PBMC_RNAseq_36_S36_L001 iPOP_MC_PBMC_RNAseq_37_S37_L001 iPOP_MC_PBMC_RNAseq_38_S38_L001 iPOP_MC_PBMC_RNAseq_39_S39_L001 iPOP_MC_PBMC_RNAseq_4_S4_L001 iPOP_MC_PBMC_RNAseq_40_S40_L001 iPOP_MC_PBMC_RNAseq_41_S41_L001 iPOP_MC_PBMC_RNAseq_42_S42_L001 iPOP_MC_PBMC_RNAseq_43_S43_L001 iPOP_MC_PBMC_RNAseq_44_S44_L001 iPOP_MC_PBMC_RNAseq_45_S45_L001 iPOP_MC_PBMC_RNAseq_46_S46_L001 iPOP_MC_PBMC_RNAseq_47_S47_L001 iPOP_MC_PBMC_RNAseq_5_S5_L001 iPOP_MC_PBMC_RNAseq_6_S6_L001 iPOP_MC_PBMC_RNAseq_7_S7_L001 iPOP_MC_PBMC_RNAseq_8_S8_L001 iPOP_MC_PBMC_RNAseq_9_S9_L001 Undetermined_S0_L001
do
samtools view -S &quot;$i&quot;.sam -b &gt; &quot;$i&quot;.bam; done</code></pre>
</div>
